src/storage/src/render/upsert/types.rs

```diff
@@ -7,85 +7,478 @@
+    ///
+    /// We are interested in extracting the cur_value of each key and discard all prev_values
+    /// that might be included in the stream. Since the history of prev_values always comes in
+    /// pairs, computing the XOR of those is always going to cancel their effects out. Also,
+    /// since XOR is commutative this property is true independent of the order. The same is
+    /// true for the summations of the length and checksum since the sum will contain the
+    /// unrelated values zero times.
+    ///
+    /// Therefore the accumulators will end up precisely in one of two states:
+    /// 1. diff == 0, checksum == 0, value == [0..] => the key is not present
+    /// 2. diff == 1, checksum == checksum(cur_value) value == cur_value => the key is present
+    ///
+    /// ## Robustness
+    ///
+    /// In the absense of bugs, accumulating the diff and checksum is not required since we know
+    /// that a well formed snapshot always satisfies XOR(bincode(values)) == bincode(cur_value).
+    /// However bugs may happen and so storing 16 more bytes per key to have a very high
+    /// guarantee that we're not decoding garbage is more than worth it.
+    /// The main key->value used to store previous values.
+    #[allow(clippy::as_conversions)]
+    pub fn merge_update(
+        &mut self,
+        value: UpsertValue,
+        diff: mz_repr::Diff,
+        bincode_opts: BincodeOpts,
+        bincode_buffer: &mut Vec<u8>,
+    ) -> bool {
+        if let Self::Snapshotting(Snapshotting {
+            value_xor,
+            len_sum,
+            checksum_sum,
+            diff_sum,
+        }) = self
+        {
+            bincode_buffer.clear();
+            bincode_opts
+                .serialize_into(&mut *bincode_buffer, &value)
+                .unwrap();
+            let len = i64::try_from(bincode_buffer.len()).unwrap();
+
+            *diff_sum += diff;
+            *len_sum += len.wrapping_mul(diff);
+            // Truncation is fine (using `as`) as this is just a checksum
+            *checksum_sum += (seahash::hash(&*bincode_buffer) as i64).wrapping_mul(diff);
+
+            // XOR of even diffs cancel out, so we only do it if diff is odd
+            if diff.abs() % 2 == 1 {
+                if value_xor.len() < bincode_buffer.len() {
+                    value_xor.resize(bincode_buffer.len(), 0);
+                }
+                for (acc, val) in value_xor.iter_mut().zip(bincode_buffer.drain(..)) {
+                    *acc ^= val;
+                }
+            }
+
+            return diff_sum.0 == 0 && checksum_sum.0 == 0 && value_xor.iter().all(|&x| x == 0);
+        } else {
+            panic!("`merge_update` called after snapshot consolidation")
+        }
+    }
+
+    /// After consolidation of a snapshot, we assume that all values in the `UpsertState` implementation
+    /// are `Self::Snapshotting`, with a `diff_sum` of 1. Afterwards, if we need to retract one of
```

Comment:
This comment is slightly inaccurate since the implementation below also allows `diff_sum` to be `0` provided that everything else is zero.

Reply:
Technically we will delete all the `0`'s as part of `merge_snapshot_chunk`, but this method doesnt know that, so ill fix the comment